pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart

```diff
@@ -510,9 +505,15 @@ ${e.message}
         ''');
             return null;
           }
+
           final outdatedFile =
               await dependenciesHashes.findOutdatedFileSystemEntity();
-          if (outdatedFile == null) {
+          final environmentChanged = (!await environmentFile.exists()) ||
```

Comment:
For files we use hashes to determine whether they changed. But for environment we use special logic here (based on equality). Why not rely on one mechanism for change detection? (the whole assumption is any input changes then md5/sha256 changes)

So I'd factor this into the dependencies file:
* When creating it we give it all the files it should use for hashing and all environment vars it should use for hashing.
* When we check whether something changed, we give it the current environment.
* Then the entire change detection logic is encapsulated in one place

Then we wouldn't need yet another file (`environment.json`) either. All things that influence caching would be encapsulated into one place.

It would be something like this:
```dart
class DependenciesHashFile {
  final File _file;
  FileSystemHashes _hashes;
  ContentHashes _contentHashes;
}

// Maps files paths to content hash
class FileSystemHashes {
  final List<FilesystemEntityHash> files;
}

// Maps abstract keys to value hash
class ContentHashes {
  final Map<String, String> contentHashes;
}
```

This `ContentHashes` could be not only used for environments but other things (e.g. in last review I mentioned one could have `<dart-sdk-path>` as symbolic key in there)

An added benefit is that we don't store the environment values (which may even contain secrets), which is a plus IMHO.

Reply:
That's a nice idea. 👍 

We'll need to refactor a bit how the logging is done for if changes are detected. If we want to give informative log messages, we should probably save per environment key (instead of hashing the whole env), and make it `EnvironmentHashes` instead of making it abstract.